Change DSet.Fold to use state separately per Parition.

jinlmsft opened the pull request:
The pull request address issue #12.

Change DSet.Fold to use state separately deserialized per partition. A FoldWithCommonStatePerNode is retained for internal implementation that takes advantage of common state per node.

zhuweirong commented:
Should not include the change to 
- docs/content/index.fsx
- docs/tools/templates/template.cshtml

Should include a unit test that verifies the behavior of fold.  For example, we can have a test that has 4 partitions on two nodes. The elements in the partitions all have value 0. Then we have "State" as 1.  Both folder and aggregator is "+". Then the expected result should be "4 + 2 = 6".  Before the fix, the result would be "2 + 2 = 4".

zhuweirong commented:
Please try to squash all commits related to a single issue into a single commit. See: https://github.com/ginatrapani/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

wezhu commented:
test

Commit:
DSet.fold use separate state per partition.

Change DSet.Fold so that the default API will provide a state separately deserialized for each partition. An internal function FoldWithCommonStatePerNode is retained for internal
implementation that can take advantage of common state per node. A UnitTest has been added to validate the implementation.
